core/store/src/trie/mem/updating.rs

```diff
@@ -33,15 +71,104 @@ pub enum UpdatedMemTrieNode {
+        child: GenericNodeOrIndex<GenericTrieNodePtr>,
     },
     /// Corresponds to either a Branch or BranchWithValue node.
     Branch {
-        children: Box<[Option<OldOrUpdatedNodeId>; 16]>,
+        children: Box<[Option<GenericNodeOrIndex<GenericTrieNodePtr>>; 16]>,
         value: Option<FlatStateValue>,
     },
 }
 
+pub type UpdatedMemTrieNode = GenericUpdatedTrieNode<MemTrieNodeId, FlatStateValue>;
+
+/// An updated node with its memory usage.
+/// Needed to recompute subtree function (memory usage) on the fly.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GenericUpdatedTrieNodeWithSize<GenericTrieNodePtr, FlatStateValue> {
+    pub node: GenericUpdatedTrieNode<GenericTrieNodePtr, FlatStateValue>,
+    pub memory_usage: u64,
+}
+
+impl<N, V> GenericUpdatedTrieNodeWithSize<N, V> {
+    pub fn empty() -> Self {
+        Self { node: GenericUpdatedTrieNode::Empty, memory_usage: 0 }
+    }
+}
+
+impl<GenericTrieNodePtr, FlatStateValue: HasLength>
+    GenericUpdatedTrieNode<GenericTrieNodePtr, FlatStateValue>
+{
+    fn memory_usage_value(value_length: u64) -> u64 {
+        value_length * TRIE_COSTS.byte_of_value + TRIE_COSTS.node_cost
+    }
+
+    /// Returns the memory usage of the **single** node, in Near's trie cost
+    /// terms, not in terms of the physical memory usage.
+    /// TODO(#12324): replace `TrieNode::memory_usage_direct_internal` and
+    ///`MemTrieNodeView::memory_usage` by this method.
+    pub fn memory_usage_direct(&self) -> u64 {
+        match self {
+            Self::Empty => {
+                // DEVNOTE: empty nodes don't exist in storage.
+                // In the in-memory implementation Some(TrieNode::Empty) and None are interchangeable as
+                // children of branch nodes which means cost has to be 0
+                0
+            }
+            Self::Leaf { extension, value } => {
+                TRIE_COSTS.node_cost
+                    + (extension.len() as u64) * TRIE_COSTS.byte_of_key
+                    + Self::memory_usage_value(value.len())
+            }
+            Self::Branch { value, .. } => {
+                TRIE_COSTS.node_cost
+                    + value.as_ref().map_or(0, |value| Self::memory_usage_value(value.len()))
+            }
+            Self::Extension { extension, .. } => {
+                TRIE_COSTS.node_cost + (extension.len() as u64) * TRIE_COSTS.byte_of_key
+            }
+        }
+    }
+}
+
+pub type UpdatedMemTrieNodeWithSize = GenericUpdatedTrieNodeWithSize<MemTrieNodeId, FlatStateValue>;
+
+/// Trait for trie updates to handle updated nodes.
+pub(crate) trait GenericTrieUpdate<'a, GenericTrieNodePtr, FlatStateValue> {
+    /// If the ID was old, converts underlying node to an updated one.
+    fn generic_ensure_updated(
+        &mut self,
+        node: GenericNodeOrIndex<GenericTrieNodePtr>,
+    ) -> Result<GenericUpdatedNodeId, StorageError>;
+
+    /// Takes a node from the set of updated nodes, setting it to None.
```

Comment:
nit: It might help readers to add a high level summary of the slot mechanism to the trait's documentation. This stuff itself may not be so confusing but when I was reading the slot mechanism for the first time before, I was very confused when I also had to understand a bunch of other things in the update logic.